File: src/prod/src/data/txnreplicator/logrecordlib/CopyMetadata.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ktl;
using namespace Data::LogRecordLib;
using namespace TxnReplicator;
using namespace Data::Utilities;

CopyMetadata::CopyMetadata(
    __in ULONG32 copyStateMetadataVersion,
    __in ProgressVector & progressVector,
    __in Epoch & startingEpoch,
    __in LONG64 startingLogicalSequenceNumber,
    __in LONG64 checkpointLsn,
    __in LONG64 uptoLsn,
    __in LONG64 highestStateProviderCopiedLsn)
    : copyStateMetadataVersion_(copyStateMetadataVersion)
    , progressVector_(&progressVector)
    , startingEpoch_(startingEpoch)
    , startingLogicalSequenceNumber_(startingLogicalSequenceNumber)
    , checkpointLsn_(checkpointLsn)
    , uptoLsn_(uptoLsn)
    , highestStateProviderCopiedLsn_(highestStateProviderCopiedLsn)
{
}

CopyMetadata::~CopyMetadata()
{
}

CopyMetadata::SPtr CopyMetadata::Create(
    __in ULONG32 copyStateMetadataVersion,
    __in ProgressVector & progressVector,
    __in Epoch & startingEpoch,
    __in LONG64 startingLogicalSequenceNumber,
    __in LONG64 checkpointLsn,
    __in LONG64 uptoLsn,
    __in LONG64 highestStateProviderCopiedLsn,
    __in KAllocator & allocator)
{
    CopyMetadata * pointer = _new(COPYMETADATA_TAG, allocator)CopyMetadata(
        copyStateMetadataVersion, 
        progressVector, 
        startingEpoch, 
        startingLogicalSequenceNumber, 
        checkpointLsn, 
        uptoLsn, 
        highestStateProviderCopiedLsn);

    ASSERT_IFNOT(
        pointer != nullptr,
        "Failed to create CopyMetadata");

    return CopyMetadata::SPtr(pointer);
}

CopyMetadata::SPtr CopyMetadata::Deserialize(
    __in OperationData const & operationData,
    __in KAllocator & allocator)
{
    ProgressVector::SPtr copiedProgressVector = ProgressVector::Create(allocator);

    KBuffer::CSPtr buffer = operationData[0];
    BinaryReader br(*buffer, allocator);

    ULONG32 copyStateMetadataVersion;
    LONG64 startingLogicalSequenceNumber;
    LONG64 checkpointLsn;
    LONG64 uptoLsn;
    LONG64 highestStateProviderCopiedLsn;

    LONG64 epochDataLossVersion;
    LONG64 epochConfigurationVersion;

    br.Read(copyStateMetadataVersion);
    copiedProgressVector->Read(br, false);
    
    br.Read(epochDataLossVersion);
    br.Read(epochConfigurationVersion);
    Epoch startingEpoch(epochDataLossVersion, epochConfigurationVersion);

    br.Read(startingLogicalSequenceNumber);
    br.Read(checkpointLsn);
    br.Read(uptoLsn);
    br.Read(highestStateProviderCopiedLsn);

    ASSERT_IFNOT(
        copyStateMetadataVersion != 1 || buffer->QuerySize() == br.Position,
        "Invalid copy metadata during deserialization, version: {0} buffer size: {1} br.Position: {2}", 
        copyStateMetadataVersion,
        buffer->QuerySize(),
        br.Position);

    return Create(
        copyStateMetadataVersion, 
        *copiedProgressVector, 
        startingEpoch, 
        startingLogicalSequenceNumber, 
        checkpointLsn, 
        uptoLsn, 
        highestStateProviderCopiedLsn, 
        allocator);
}

KBuffer::SPtr CopyMetadata::Serialize(
    __in CopyMetadata const & copyMetadata,
    __in KAllocator & allocator)
{
    BinaryWriter bw(allocator);

    bw.Write(copyMetadata.copyStateMetadataVersion_);
    copyMetadata.progressVector_->Write(bw);
    bw.Write(copyMetadata.startingEpoch_.DataLossVersion);
    bw.Write(copyMetadata.startingEpoch_.ConfigurationVersion);
    bw.Write(copyMetadata.startingLogicalSequenceNumber_);
    bw.Write(copyMetadata.checkpointLsn_);
    bw.Write(copyMetadata.uptoLsn_);
    bw.Write(copyMetadata.highestStateProviderCopiedLsn_);

    KBuffer::SPtr buffer;
    NTSTATUS status = KBuffer::Create(bw.Position, buffer, allocator);

    ASSERT_IFNOT(
        status == STATUS_SUCCESS,
        "CopyMetadata::Serialize failed to allocate KBuffer {0:x}",
        status);

    buffer = bw.GetBuffer(0);

    return buffer;
}

File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/CopyMetadata.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System.Fabric;
    using System.IO;

    /// <summary>
    /// The copy metadata.
    /// </summary>
    internal class CopyMetadata
    {
        /// <summary>
        /// The copy state metadata version.
        /// </summary>
        private readonly int copyStateMetadataVersion;

        /// <summary>
        /// The progress vector.
        /// </summary>
        private readonly ProgressVector progressVector;

        /// <summary>
        /// The starting epoch.
        /// </summary>
        private readonly Epoch startingEpoch;

        /// <summary>
        /// The starting logical sequence number.
        /// </summary>
        private readonly LogicalSequenceNumber startingLogicalSequenceNumber;

        /// <summary>
        /// The checkpoint lsn.
        /// </summary>
        private readonly LogicalSequenceNumber checkpointLsn;

        /// <summary>
        /// The upto lsn.
        /// </summary>
        private readonly LogicalSequenceNumber uptoLsn;

        /// <summary>
        /// The highest state provider copied lsn.
        /// </summary>
        private readonly LogicalSequenceNumber highestStateProviderCopiedLsn;

        /// <summary>
        /// Gets the highest state provider copied lsn.
        /// </summary>
        internal LogicalSequenceNumber HighestStateProviderCopiedLsn
        {
            get
            {
                return this.highestStateProviderCopiedLsn;
            }
        }

        /// <summary>
        /// Gets the upto lsn.
        /// </summary>
        public LogicalSequenceNumber UptoLsn
        {
            get
            {
                return this.uptoLsn;
            }
        }

        /// <summary>
        /// Gets the checkpoint lsn.
        /// </summary>
        public LogicalSequenceNumber CheckpointLsn
        {
            get
            {
                return this.checkpointLsn;
            }
        }

        /// <summary>
        /// Gets the starting logical sequence number.
        /// </summary>
        public LogicalSequenceNumber StartingLogicalSequenceNumber
        {
            get
            {
                return this.startingLogicalSequenceNumber;
            }
        }

        /// <summary>
        /// Gets the starting epoch.
        /// </summary>
        public Epoch StartingEpoch
        {
            get
            {
                return this.startingEpoch;
            }
        }

        /// <summary>
        /// Gets the progress vector.
        /// </summary>
        public ProgressVector ProgressVector
        {
            get
            {
                return this.progressVector;
            }
        }

        /// <summary>
        /// Gets the copy state metadata version.
        /// </summary>
        public int CopyStateMetadataVersion
        {
            get
            {
                return this.copyStateMetadataVersion;
            }
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="CopyMetadata"/> class.
        /// </summary>
        /// <param name="copyStateMetadataVersion">
        /// The copy state metadata version.
        /// </param>
        /// <param name="progressVector">
        /// The progress vector.
        /// </param>
        /// <param name="startingEpoch">
        /// The starting epoch.
        /// </param>
        /// <param name="startingLogicalSequenceNumber">
        /// The starting logical sequence number.
        /// </param>
        /// <param name="checkpointLsn">
        /// The checkpoint lsn.
        /// </param>
        /// <param name="uptoLsn">
        /// The upto lsn.
        /// </param>
        /// <param name="highestStateProviderCopiedLsn">
        /// The highest state provider copied lsn.
        /// </param>
        public CopyMetadata(
            int copyStateMetadataVersion, 
            ProgressVector progressVector, 
            Epoch startingEpoch, 
            LogicalSequenceNumber startingLogicalSequenceNumber, 
            LogicalSequenceNumber checkpointLsn, 
            LogicalSequenceNumber uptoLsn, 
            LogicalSequenceNumber highestStateProviderCopiedLsn)
        {
            this.copyStateMetadataVersion = copyStateMetadataVersion;
            this.progressVector = progressVector;
            this.startingEpoch = startingEpoch;
            this.startingLogicalSequenceNumber = startingLogicalSequenceNumber;
            this.checkpointLsn = checkpointLsn;
            this.uptoLsn = uptoLsn;
            this.highestStateProviderCopiedLsn = highestStateProviderCopiedLsn;
        }

        /// <summary>
        /// The read from operation data.
        /// </summary>
        /// <param name="operationData">
        /// The operation data.
        /// </param>
        /// <returns>
        /// The <see cref="CopyMetadata"/>.
        /// </returns>
        public static CopyMetadata ReadFromOperationData(OperationData operationData)
        {
            var copiedProgressVector = new ProgressVector();

            using (var br = new BinaryReader(new MemoryStream(operationData[0].Array, operationData[0].Offset, operationData[0].Count)))
            {
                var copyStateMetadataVersion = br.ReadInt32();
                copiedProgressVector.Read(br, false);
                var startingEpoch = new Epoch(br.ReadInt64(), br.ReadInt64());
                var startingLsn = new LogicalSequenceNumber(br.ReadInt64());
                var checkpointLsn = new LogicalSequenceNumber(br.ReadInt64());
                var uptoLsn = new LogicalSequenceNumber(br.ReadInt64());
                var highestStateProviderCopiedLsn = new LogicalSequenceNumber(br.ReadInt64());

                // Note that if version is 1, then the size must be exactly as expected.
                // Else the rest of the data is expected to be not required.
                Utility.Assert(
                    copyStateMetadataVersion != 1 || operationData[0].Array.Length == br.BaseStream.Position, 
                    "Unexpected copy state metadata size. Version {0} logMetadata.Length {1} Position {2}", 
                    copyStateMetadataVersion, operationData[0].Array.Length, br.BaseStream.Position);

                return new CopyMetadata(
                    copyStateMetadataVersion, 
                    copiedProgressVector, 
                    startingEpoch, 
                    startingLsn, 
                    checkpointLsn, 
                    uptoLsn, 
                    highestStateProviderCopiedLsn);
            }
        }
    }
}